Issue #12695 - don't return ByteBuffer from .toMappedBuffer if FS reports as UnsupportedOperationException.

joakime opened the pull request:
To fix issue with UnsupportedOperationException, just have toMappedBuffer return null (that is an existing contract) if that exception is triggered.

joakime committed:
Issue #12695 - don't return ByteBuffer from .toMappedBuffer if FS reports as UnsupportedOperationException.

joakime commented:
This is an incomplete fix, as the usages of `toMappedBuffer` in `org.eclipse.jetty.http.content.FileMappingHttpContentFactory` incorrectly assume that they will always get a mapped ByteBuffer.

joakime committed:
Issue #12695 - reject attempts to memory map files from FileSystems that do not support it.

joakime committed:
Issue #12695 - restore FileBufferedResponseHandlerTest

+ Fallback to normal filesystem read if memory mapping is not supported.

joakime commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpTester.java:
```diff
@@ -262,17 +262,25 @@ public static Response parseResponse(Input input) throws IOException
         return parseResponse(input, false);
     }
 
+    public static Response parseResponse(Input input, Response response) throws IOException
```
While restoring `FileBufferedResponseHandlerTest` (from earlier versions of Jetty) this method in `HttpTester` was also restored.

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -205,9 +206,9 @@ protected Action process() throws Exception
+                    long len = Math.min(MAX_BUFFER_SIZE, fileLength - _pos);
                     _last = (_pos + len == fileLength);
-                    ByteBuffer buffer = BufferUtil.toMappedBuffer(_filePath, _pos, len);
+                    ByteBuffer buffer = readByteBuffer(_filePath, _pos, len);
```
`readByteBuffer()` could allocate a 1 GB buffer, each time a file is served! Yikes!

joakime replied:
If you have 1 GB of generated response body, then yup, and that's what the javadoc at the top of this handler warns about.

joakime replied:
Addressed this with the new internal `SendFileCallback` class, along with bufferSize configurable (defaults to 64kb)

joakime replied:
No longer relevant.

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -229,5 +230,23 @@ protected void onFailure(Throwable cause)
+            try (SeekableByteChannel channel = Files.newByteChannel(path))
+            {
+                channel.position(pos);
+                buffer = ByteBuffer.allocate((int)len);
```
Please use direct byte buffers by default.

joakime replied:
Addressed this with the new internal `SendFileCallback` class

joakime replied:
No longer relevant.

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -229,5 +230,23 @@ protected void onFailure(Throwable cause)
+            {
+                channel.position(pos);
+                buffer = ByteBuffer.allocate((int)len);
+                channel.read(buffer);
```
IIRC `channel.read()` does not guarantee that the buffer will be filled as this API is non-blocking. You are going to either spin loop or use some form of blocking API instead.

joakime replied:
I don't see that as a problem, we read what we can and return, even if the buffer isn't full, the iterating calback that calls this method will just get a new position to read from, right?

joakime replied:
The new `FileBufferedResponseHandler.setUseFileMapping(boolean)` allows me to test these new features in unit tests too.

joakime replied:
No longer relevant.

joakime committed:
Attempt to answer review questions

joakime committed:
Issue #12695 - FileBufferedResponseHandler.useFileMapping

+ Add new .setUseFileMapping(boolean) flag
  (defaults to true)
+ Add test case variations for non-mapping
+ Fix position calculation

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -193,41 +215,154 @@ private void commit(Callback callback)
+
+        enum Mode
+        {
+            DISCOVER,
```
`AUTO` sounds better than `DISCOVER` to me as it implies that some logic is going to automatically choose between the other two.

joakime replied:
No longer relevant, using `Content.copy()` now

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -193,41 +215,154 @@ private void commit(Callback callback)
+            return buffer;
+        }
+
+        private ByteBuffer toRead(Path path, long pos) throws IOException
```
This method allocates a 64 KB direct buffer, opens and closes a file for each 64 KB chunk of data in the file, which is going to be way too costly.

The buffer should be allocated at most once and the file should be opened and closed at most once.

joakime replied:
No longer relevant, using `Content.copy()` now

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -193,41 +215,154 @@ private void commit(Callback callback)
+        }
+    }
+
+    // TODO: can this be made generic enough to put into jetty-io somewhere?
```
To answer your comment: I don't think so. We do 99% of our IO's using `Content.Source` and `Content.Sink` for which we already have `Content.copy()`.

joakime replied:
Implemented `Content.copy()` technique instead.

lorban reviewed (changes requested):
I'm wondering if this whole implementation could be replaced with a simple `Content.copy()`?

Writing a `Content.Sink` that does `_interceptor.getNextInterceptor().write(buffer, last, callback)` looks trivial; so if a memory-mapping `Content.Source` could be created, with a fallback to `Content.Source.from(Path)` if mmap is disabled or unavailable could be written then the whole `Content` logic could be re-used.

joakime committed:
Issue #12695 - Cleanup FileBufferedResponseHandler

+ Remove useFileMapping config
+ Remove custom IteratingCallback
+ Use Content.Source.from(Path)
+ New InterceptorSink

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferPool.java:
```diff
@@ -122,6 +122,21 @@ class Sized extends Wrapper
+         * @param size The specified size in bytes of the buffer, or -1 for a default
+         * @return the pool as a sized pool
+         */
+        public static ByteBufferPool.Sized as(ByteBufferPool pool, boolean direct, int size)
```
I don't think I like this new helper as it takes parameters that could be ignored without any hint.

joakime replied:
Removed

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -193,41 +204,27 @@ private void commit(Callback callback)
-                private final long fileLength = _filePath.toFile().length();
-                private long _pos = 0;
-                private boolean _last = false;
+            ByteBufferPool.Sized sizedPool = ByteBufferPool.Sized.as(getServer().getByteBufferPool(), true, getBufferSize());
```
I think I'd rather see a new instance of `ByteBufferPool.Sized` being created.

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -193,41 +204,27 @@ private void commit(Callback callback)
-                    _pos += len;
-                    return Action.SCHEDULED;
-                }
+    private static class InterceptorSink implements Content.Sink
```
You could inline this class as a simple lambda:
```java
Content.Sink sink = (last, byteBuffer, callback) -> _interceptor.write(last, byteBuffer, callback);
```

lorban commented on jetty-ee9/jetty-ee9-nested/src/main/java/org/eclipse/jetty/ee9/nested/FileBufferedResponseHandler.java:
```diff
@@ -193,41 +204,27 @@ private void commit(Callback callback)
-                private long _pos = 0;
-                private boolean _last = false;
+            ByteBufferPool.Sized sizedPool = ByteBufferPool.Sized.as(getServer().getByteBufferPool(), true, getBufferSize());
+            Content.Source source = Content.Source.from(sizedPool, _filePath);
```
This does not `mmap` files naymore, but honestly I think it is for the best.

joakime replied:
Supporting mmap from Content.Source.from(path) is still possible, but in a future PR.

lorban reviewed (approved):
I'll leave it up to you to decide to keep the new `ByteBufferPool.Sized.as()` helper or not, as well as inlining the `InterceptorSink` or not.

LGTM either way.

joakime committed:
Changes from review
